src/node_io.h

```diff
@@ -0,0 +1,236 @@
+                        int status,
+                        size_t length);
+
+typedef void (*pull_cb)(io_pull_t* handle,
+                        int status,
+                        io_buf_t* bufs,
+                        size_t count,
+                        size_t length,
+                        io_push_t* source);
+
+typedef void (*alloc_cb)(io_bind_t* handle,
+                         size_t length,
+                         io_buf_t** bufs,
+                         size_t* count);
+
+typedef void (*free_cb)(io_bind_t* handle,
+                        size_t length,
+                        io_buf_t* bufs,
+                        size_t count);
+
+struct io_pull_s {
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+  pull_cb pull = nullptr;
+  void* user_data = nullptr;
+#endif
+};
+
+struct io_push_s {
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+  push_cb push = nullptr;
+  void* user_data = nullptr;
+#endif
+};
+
+struct io_bind_s {
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+  bind_cb bind = nullptr;
+  unbind_cb unbind = nullptr;
+  alloc_cb alloc = nullptr;
+  free_cb free = nullptr;
+  void* user_data = nullptr;
+#endif
+};
+
+// Essentially the same as a uv_buf_t, except uint8_t rather than char
+struct io_buf_s {
+  uint8_t* base;
+  size_t len;
+};
+
+
+inline void io_bind_set_user_data(io_bind_t* handle, void* user_data);
+inline void io_pull_set_user_data(io_pull_t* handle, void* user_data);
+inline void io_push_set_user_data(io_push_t* handle, void* user_data);
+inline void* io_bind_get_user_data(io_bind_t* handle);
+inline void* io_pull_get_user_data(io_pull_t* handle);
+inline void* io_push_get_user_data(io_push_t* handle);
+
+inline void io_bind_set_alloc_cb(io_bind_t* handle, alloc_cb cb);
+inline void io_bind_set_free_cb(io_bind_t* handle, free_cb cb);
+inline void io_bind_set_bind_cb(io_bind_t* handle, bind_cb cb);
+inline void io_bind_set_unbind_cb(io_bind_t* handle, unbind_cb cb);
+inline void io_pull_set_pull_cb(io_pull_t* handle, pull_cb cb);
+inline void io_push_set_push_cb(io_push_t* handle, push_cb cb);
+
+inline alloc_cb io_bind_get_alloc_cb(io_bind_t* handle);
+inline free_cb io_bind_get_free_cb(io_bind_t* handle);
+inline bind_cb io_bind_get_bind_cb(io_bind_t* handle);
+inline unbind_cb io_bind_get_unbind_cb(io_bind_t* handle);
+inline pull_cb io_pull_get_pull_cb(io_pull_t* handle);
+inline push_cb io_push_get_push_cb(io_push_t* handle);
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+inline void io_bind_set_source(io_bind_t* handle, Source* source);
+inline void io_bind_set_sink(io_bind_t* handle, Sink* sink);
+#endif
+
+inline Source* io_bind_get_source(io_bind_t* handle);
+inline Sink* io_bind_get_sink(io_bind_t* handle);
+
+inline int io_get_error(int status);
+
+class Source {
+ public:
+  Source(Environment* env) : env_(env) {}
+  virtual ~Source() {}
+
+  virtual int Pull(io_pull_t* handle,
+                   size_t* length = 0,
```

Comment:
Is this `length` written to? (Guaranteed to be?)

What does the passed-in length do? Se the maximum?